File: MinStack_Problem2.java
// Time Complexity : push - O(1), pop - O(1), top, getMin - O(1)
// Space Complexity : O(1) - As we are using same stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

class MinStack {

    private Stack<Integer> stack;
    private Integer min;
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack();
        min = Integer.MAX_VALUE;
    }
    
    public void push(int x) {
        if(min >= x){
            stack.push(min);
            min = x;
        }
        stack.push(x);
    }
    
    public void pop() {
        int popped = stack.pop();
        if(popped == min){
            min = stack.pop();
        }
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Min_Stack_Problem2.py
# Time Complexity: O(1)     during all push, pop, getmin
# Space Complexity: O(n), worst case O(2n)
# Did this code successfully run on Leetcode: Yes

import sys


class MinStack:

    def __init__(self):
        self.min = sys.maxsize
        self.stack = []

    def push(self, val: int) -> None:
        if val <= self.min:
            self.stack.append(self.min)
            self.stack.append(val)
            self.min = val
        else:
            self.stack.append(val)

    def pop(self) -> None:
        if self.stack[-1] == self.min:
            temp = self.stack.pop()
            self.min = self.stack.pop()
            return temp
        else:
            self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min


minStack = MinStack()
minStack.push(0)
minStack.push(1)
minStack.push(0)
minStack.getMin()
minStack.pop()
minStack.getMin()
minStack.top()
minStack.getMin()